File: app/src/main/java/wangdaye/com/geometricweather/common/utils/helpers/ShortcutsHelper.java
package wangdaye.com.geometricweather.common.utils.helpers;

import android.content.Context;
import android.content.pm.ShortcutInfo;
import android.content.pm.ShortcutManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.Icon;
import android.os.Build;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.core.content.ContextCompat;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.common.basic.models.Location;
import wangdaye.com.geometricweather.common.basic.models.weather.Weather;
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode;
import wangdaye.com.geometricweather.db.DatabaseHelper;
import wangdaye.com.geometricweather.theme.resource.ResourceHelper;
import wangdaye.com.geometricweather.theme.resource.providers.ResourceProvider;
import wangdaye.com.geometricweather.theme.resource.ResourcesProviderFactory;

/**
 * Shortcuts manager.
 * */

@RequiresApi(api = Build.VERSION_CODES.N_MR1)
public class ShortcutsHelper {

    public static void refreshShortcutsInNewThread(final Context c, List<Location> locationList) {
        AsyncHelper.runOnIO(() -> {
            ShortcutManager shortcutManager = c.getSystemService(ShortcutManager.class);
            if (shortcutManager == null) {
                return;
            }

            List<Location> list = Location.excludeInvalidResidentLocation(c, locationList);
            ResourceProvider provider = ResourcesProviderFactory.getNewInstance();
            List<ShortcutInfo> shortcutList = new ArrayList<>();

            // refresh button.
            Icon icon;
            String title = c.getString(R.string.refresh);
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                icon = Icon.createWithAdaptiveBitmap(
                        drawableToBitmap(
                                Objects.requireNonNull(
                                        ContextCompat.getDrawable(c, R.drawable.shortcuts_refresh_foreground)
                                )
                        )
                );
            } else {
                icon = Icon.createWithResource(c, R.drawable.shortcuts_refresh);
            }
            shortcutList.add(
                    new ShortcutInfo.Builder(c, "refresh_data")
                            .setIcon(icon)
                            .setShortLabel(title)
                            .setLongLabel(title)
                            .setIntent(IntentHelper.buildAwakeUpdateActivityIntent())
                            .build()
            );

            // location list.
            int count = Math.min(
                    shortcutManager.getMaxShortcutCountPerActivity() - 1,
                    list.size()
            );
            for (int i = 0; i < count; i ++) {
                Weather weather = DatabaseHelper.getInstance(c).readWeather(list.get(i));
                if (weather != null) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        icon = getAdaptiveIcon(
                                provider,
                                weather.getCurrent().getWeatherCode(),
                                list.get(i).isDaylight()
                        );
                    } else {
                        icon = getIcon(
                                provider,
                                weather.getCurrent().getWeatherCode(),
                                list.get(i).isDaylight()
                        );
                    }
                } else {
                    icon = getIcon(provider, WeatherCode.CLEAR, true);
                }

                title = list.get(i).isCurrentPosition() ? c.getString(R.string.current_location) : list.get(i).getCityName(c);

                shortcutList.add(
                        new ShortcutInfo.Builder(c, list.get(i).getFormattedId())
                                .setIcon(icon)
                                .setShortLabel(title)
                                .setLongLabel(title)
                                .setIntent(IntentHelper.buildMainActivityIntent(list.get(i)))
                                .build()
                );
            }

            try {
                shortcutManager.setDynamicShortcuts(shortcutList);
            } catch (Exception ignore) {
                // do nothing.
            }
        });
    }

    @NonNull
    private static Bitmap drawableToBitmap(@NonNull Drawable drawable) {
        Bitmap bitmap = Bitmap.createBitmap(
                drawable.getIntrinsicWidth(),
                drawable.getIntrinsicHeight(),
                Bitmap.Config.ARGB_8888
        );

        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, drawable.getIntrinsicWidth(), drawable.getIntrinsicHeight());
        drawable.draw(canvas);

        return bitmap;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    @NonNull
    private static Icon getAdaptiveIcon(ResourceProvider provider, WeatherCode code, boolean daytime) {
        return Icon.createWithAdaptiveBitmap(
                drawableToBitmap(
                        ResourceHelper.getShortcutsForegroundIcon(provider, code, daytime)
                )
        );
    }

    @NonNull
    private static Icon getIcon(ResourceProvider provider, WeatherCode code, boolean daytime) {
        return Icon.createWithBitmap(
                drawableToBitmap(
                        ResourceHelper.getShortcutsIcon(provider, code, daytime)
                )
        );
    }
}

File: app/src/main/java/wangdaye/com/geometricweather/common/utils/helpers/ShortcutsHelper.kt
package wangdaye.com.geometricweather.common.utils.helpers

import android.content.Context
import android.content.pm.ShortcutInfo
import android.content.pm.ShortcutManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.drawable.Drawable
import android.graphics.drawable.Icon
import android.os.Build
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import wangdaye.com.geometricweather.R
import wangdaye.com.geometricweather.common.basic.models.Location
import wangdaye.com.geometricweather.common.basic.models.weather.WeatherCode
import wangdaye.com.geometricweather.db.DatabaseHelper
import wangdaye.com.geometricweather.theme.resource.ResourceHelper
import wangdaye.com.geometricweather.theme.resource.ResourcesProviderFactory
import wangdaye.com.geometricweather.theme.resource.providers.ResourceProvider

/**
 * Shortcuts manager.
 */
@RequiresApi(api = Build.VERSION_CODES.N_MR1)
object ShortcutsHelper {

    @JvmStatic
    fun refreshShortcutsInNewThread(c: Context, locationList: List<Location>) {
        AsyncHelper.runOnIO {
            val shortcutManager = c.getSystemService(ShortcutManager::class.java) ?: return@runOnIO
            val list = Location.excludeInvalidResidentLocation(c, locationList)
            val provider = ResourcesProviderFactory.getNewInstance()
            val shortcutList = ArrayList<ShortcutInfo>()

            var icon: Icon
            var title = c.getString(R.string.refresh)
            icon = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                Icon.createWithAdaptiveBitmap(
                    drawableToBitmap(
                        ContextCompat.getDrawable(c, R.drawable.shortcuts_refresh_foreground)!!
                    )
                )
            } else {
                Icon.createWithResource(c, R.drawable.shortcuts_refresh)
            }
            shortcutList.add(
                ShortcutInfo.Builder(c, "refresh_data")
                    .setIcon(icon)
                    .setShortLabel(title)
                    .setLongLabel(title)
                    .setIntent(IntentHelper.buildAwakeUpdateActivityIntent())
                    .build()
            )

            val count = (shortcutManager.maxShortcutCountPerActivity - 1).coerceAtMost(list.size)
            for (i in 0 until count) {
                val weather = DatabaseHelper.getInstance(c).readWeather(list[i])
                icon = if (weather != null) {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                        getAdaptiveIcon(
                            provider,
                            weather.current.weatherCode,
                            list[i].isDaylight
                        )
                    } else {
                        getIcon(
                            provider,
                            weather.current.weatherCode,
                            list[i].isDaylight
                        )
                    }
                } else {
                    getIcon(provider, WeatherCode.CLEAR, true)
                }
                title = if (list[i].isCurrentPosition) c.getString(R.string.current_location) else list[i].getCityName(c)
                shortcutList.add(
                    ShortcutInfo.Builder(c, list[i].formattedId)
                        .setIcon(icon)
                        .setShortLabel(title)
                        .setLongLabel(title)
                        .setIntent(IntentHelper.buildMainActivityIntent(list[i]))
                        .build()
                )
            }
            try {
                shortcutManager.dynamicShortcuts = shortcutList
            } catch (_: Exception) {
            }
        }
    }

    private fun drawableToBitmap(drawable: Drawable): Bitmap {
        val bitmap = Bitmap.createBitmap(
            drawable.intrinsicWidth,
            drawable.intrinsicHeight,
            Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, drawable.intrinsicWidth, drawable.intrinsicHeight)
        drawable.draw(canvas)
        return bitmap
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    private fun getAdaptiveIcon(provider: ResourceProvider, code: WeatherCode, daytime: Boolean): Icon {
        return Icon.createWithAdaptiveBitmap(
            drawableToBitmap(ResourceHelper.getShortcutsForegroundIcon(provider, code, daytime))
        )
    }

    private fun getIcon(provider: ResourceProvider, code: WeatherCode, daytime: Boolean): Icon {
        return Icon.createWithBitmap(
            drawableToBitmap(ResourceHelper.getShortcutsIcon(provider, code, daytime))
        )
    }
}
